### PR TITLE
DEP PHP 8 support

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-webauthn-authenticator/issues/86

Have manually registration and verification with a yubikey

### CODE REVIEW

**Review by maxime-rainville — CHANGES_REQUESTED** [review]
My site was working with the old version. I upgraded to this PR. The first time tried logging in with my Yubi key, it fail without any useful feedback. When I looked at the request, it said: `message: "Verification failed: Requires GMP or bcmath extension."`

So it looks like web-auth/webauthn-lib v3 needs at least one of those extension to work. It seems to work fine with either of them installed.

```bash
# I managed to install both extension with this
sudo apt-get install php7.4-bcmath
sudo apt-get install php7.4-gmp
sudo phpenmod bcmath
sudo phpenmod gmp
sudo service apache2 restart

```

That's not necessarily a deal breaker, but we need to take steps to point out this problem to users:
1. We need to make sure at least one of those extension is installed on SCP and CCL
2. We need to be pretty explicit in the upgrade doc and the readme for this project
3. We need to provide some useful feedback to the user if both those extension are missing

There's already warnings if you try to use the Yubi key over HTTP or without a MFA secret. Seems to me like we should be able to piggy back over those mechanisms.

Looks like this is control here: https://github.com/silverstripe/silverstripe-webauthn-authenticator/blob/ea5da59a058d9b516142fd92570ab3c7ff10a127/src/Method.php#L81-L89

image

image

**Comment by maxime-rainville** [comment]
Looks like the error is actually coming from [here](https://github.com/fgrosse/PHPASN1/blob/20299033c35f4300eb656e7e8e88cf52d1d6694e/lib/Utility/BigInteger.php#L56-L63): 
```php
            if (function_exists('gmp_add')) {
                $ret = new BigIntegerGmp();
            }
            elseif (function_exists('bcadd')) {
                $ret = new BigIntegerBcmath();
            } else {
                throw new \RuntimeException('Requires GMP or bcmath extension.');
            }
```

In a perfect world, web-auth/webauthn-lib would have a method to tell us directly that something is missing, Unfortunately, it looks like we'll have to copy bits of that logic into our codebase.

File: client/src/lib/convert.js [hunk 1]
```diff
@@ -1,5 +1,12 @@
...
+};
 
-export const byteArrayToBase64 = byteArray =>
-  btoa(String.fromCharCode(...(new Uint8Array(byteArray))));
+export const byteArrayToBase64 = byteArray => {
+  // We specifically do not want to make the 'web safe' string replacements above
+  // doing so will break this functionality
+  const uarr = new Uint8Array(byteArray);
+  return btoa(String.fromCharCode(...uarr));
+};
```
Comment (maxime-rainville) [inline client/src/lib/convert.js:12]: Looks like two method that are pretty easy to test. Maybe write some test for them if times allows.

Or even better, find a well supported third party lib to do this and use that instead.

File: src/RegisterHandler.php [hunk 2]
```diff
@@ -259,12 +256,10 @@ protected function getCredentialCreationOptions(
             $this->getUserEntity($store->getMember()),
             random_bytes(32),
             [new PublicKeyCredentialParameters('public-key', Algorithms::COSE_ALGORITHM_ES256)],
-            40000,
-            [],
-            $this->getAuthenticatorSelectionCriteria(),
-            PublicKeyCredentialCreationOptions::ATTESTATION_CONVEYANCE_PREFERENCE_NONE,
-            new AuthenticationExtensionsClientInputs()
+            40000
```
Comment (maxime-rainville) [inline src/RegisterHandler.php:259]: Do we know what that `40000` number represent? Might be worth putting it in a constant for clarity.

**Review by maxime-rainville — COMMENTED** [review]
Did a code review and left some very minor code suggestions. 

I'm not familiar at all with the `web-auth/webauthn-lib` ... I'm presuming you followed some upgrade doc to make those changes and are following some sort of best practices?

Also, if there's someone else with a Yubi key in Wellington, maybe double check that it doesn't blindly accept any Yubi key input ... or we can deploy it to a stack and one of us attempts to impersonate the other.

File: src/RegisterHandler.php (hunk 2 above)
Comment (emteknetnz) [reply src/RegisterHandler.php:259]: $timeout - Webauthn\PublicKeyCredentialCreationOptions::construct()
presumably 40 seconds

File: client/src/lib/convert.js (hunk 1 above)
Comment (emteknetnz) [reply client/src/lib/convert.js:12]: Added tests

**Review by maxime-rainville — APPROVED** [review]
Looks good ... let's just test it on our platform to make use it works.

**Comment by maxime-rainville** [comment]
Tested it on all our platforms and it works great.

**Pull request merged**